### PR TITLE
TVA-841: Domain modelling for Gathering Provider Details

### PR DESCRIPTION
**Pull request by kostasakrivos**
### Implemented domain logic GatheringProviderDetails
This is the very domain and the logic of how financial relationships
get created or updated, and thus evaluating a status for the data
onboarding process. It does not contain any technology, involving
user-interface or infrastructure needs. It currently only focuses on
the 'Gathering Provider Details' status.

Non-obvious things:

[Testing Approach]
The testing approach might not be the most obvious, but we tried to
bring how the ACs are broken down into the code itself to reduce
back-and-forth understanding between business and engineers.

Because the status 'Gathering Provider Details' is evaluated
differently between client/provider specific financial relationships,
to increase readability and cohesion, we created a different files of
tests for each group. The groups we found are the following:
- all relationships.
- provider-specific relationships.
- client-specific relationships.

As we worked bottom-up, these 3 unique files of tests form a certain
specification. So it made sense for us to create one, which gathers all
of those tests defined in each separate file. We called that
specification, `GatheringProviderDetailsTestSpec`. This makes us able to
test that single specification against the service, which implements
the business logic. So the first thing we see in the tests source code,
is the specification, and under that package, we se all individual
tests that are grouped as mentioned above.

The specification and thus all individual test files are abstractions
that must, at least, depend on the very system under test. We tried our
best to define the tests in such a way that do not depend on
implementation details, and that can be hopefully run against a
different morph of the system under test. Because we currently follow
the Clean Architecture and Ports and Adapters architectural patterns,
where our business logic is exposed via abstractions, and not concrete
implementations; we can use that to our advantage, and create a
different morph of the system under test (polymorphism).

To exercise the specification against that system under test, we need to
implement it and give it whatever is needed for that test to run. Our
current single test is for `UpsertClientFinancialRelationshipUseCase`
against an in-memory version of the driven port (repository), that is
sufficient to pass all tests. An example of the above things mentioned,
another test could be that of `UpsertClientFinancialRelationshipUseCase`
but against an in-memory or dockerised mongodb, for when that adapter is
implemented.

[Testing Library]
We brought in the library called `hamcrest` which uses the junit api,
but offers a lot more flexibility to express assertions with less noise,
and keeping a better domain-specific language. Doing so, it brings the
`Matcher<T>` api, for which there already a couple of high-level
helpers, but we can create our own matchers.
### Removed hello world test
It was there to assert the code has at least one test running, to
exercise the CI/CD pipelines. Now that we have tests for the problem
we're solving, it's redundant.

### Added Logging

[jira ticket](https://yofiworld.atlassian.net/browse/TVA-841)

### CODE REVIEW

**Comment by ytree-admin** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign 
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=y-tree-limited%2Fms-dob-fwk).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/y-tree-limited/ms-dob-fwk/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: dob-status-core/src/main/kotlin/com/ytree/dob_fwk/status/core/domain/MicroTypes.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.ytree.dob_fwk.status.core.domain
+
+import java.util.UUID
+import java.util.Currency
+
+data class FinancialRelationshipId(val value: UUID)
+data class ClientId(val value: UUID)
+data class ProviderId(val value: UUID)
+data class BaseCurrency(val value: Currency)
```
Comment (Petlyuk-Vladimir) [inline dob-status-core/src/main/kotlin/com/ytree/dob_fwk/status/core/domain/MicroTypes.kt:9]: In that case that we use crypto currency in our system java.util.Currency not a good idea. We use Int type and use id from constants>v2.currency as identifier

Comment (kostasakrivos) [reply dob-status-core/src/main/kotlin/com/ytree/dob_fwk/status/core/domain/MicroTypes.kt]: fixed in [93bb4e0](https://github.com/y-tree-limited/ms-dob-fwk/pull/2/commits/93bb4e0da12ed4206a0fa677a70caabb586c0377)

**Review by Petlyuk-Vladimir — APPROVED** [review]

**Comment by ytree-admin** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[Petlyuk-Vladimir](https://github.com/y-tree-limited/ms-dob-fwk/pull/2#pullrequestreview-1317404368 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign 
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=y-tree-limited%2Fms-dob-fwk).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/y-tree-limited/ms-dob-fwk/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by ytree-admin** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[kostasakrivos](https://github.com/y-tree-limited/ms-dob-fwk/pull/2#issuecomment-1448084449 "Approved")*, *[Petlyuk-Vladimir](https://github.com/y-tree-limited/ms-dob-fwk/pull/2#pullrequestreview-1317404368 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign 
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=y-tree-limited%2Fms-dob-fwk).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/y-tree-limited/ms-dob-fwk/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by ytree-admin** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by: *[kostasakrivos](https://github.com/y-tree-limited/ms-dob-fwk/pull/2#issuecomment-1448084449 "Approved")*, *[Petlyuk-Vladimir](https://github.com/y-tree-limited/ms-dob-fwk/pull/2#pullrequestreview-1317404368 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=y-tree-limited%2Fms-dob-fwk).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/y-tree-limited/ms-dob-fwk/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**